### PR TITLE
Cpm 454 Add endpoint to return the list of select attributes with options count

### PR DESCRIPTION
**Pull request by amigaire**
PR EE: https://github.com/akeneo/pim-enterprise-dev/pull/12635

**Description (for Contributor and Core Developer)**

Add a query to get all select option attributes' labels and their options count

**Definition Of Done (for Core Developer only)**

- [ ] Tests
- [ ] PM Validation (Story)

### CODE REVIEW

**Review by cemorin — APPROVED** [review]

File: src/Akeneo/Pim/Structure/Bundle/Query/PublicApi/AttributeOption/Sql/SqlGetOptionsCountAndTranslationByAttribute.php [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    /**
+     * @return array example:
+     *      [
+     *          ['code' => 'attribute1', 'label' => 'The Label', 'options_count' => 5],
+     *          ...
+     *      ]
+     * @throws \Doctrine\DBAL\Driver\Exception
+     * @throws \Doctrine\DBAL\Exception
+     */
+    public function search(string $locale, int $page = 1, int $limit = self::MAX_PAGE_SIZE, string $search = null): array
```
Comment (mmetayer) [inline src/Akeneo/Pim/Structure/Bundle/Query/PublicApi/AttributeOption/Sql/SqlGetOptionsCountAndTranslationByAttribute.php:32]:
```suggestion
    public function search(string $localeCode, int $limit = self::MAX_PAGE_SIZE, int $offset = 0, ?string $search = null): array
```

Comment (mmetayer) [reply src/Akeneo/Pim/Structure/Bundle/Query/PublicApi/AttributeOption/Sql/SqlGetOptionsCountAndTranslationByAttribute.php:32]: I'd prefer if the offset was calculated by the caller

File: src/Akeneo/Pim/Structure/Bundle/Query/PublicApi/AttributeOption/Sql/SqlGetOptionsCountAndTranslationByAttribute.php [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+                'offset' => $offset,
+                'search' => strtolower($search ?? '') . '%',
+                'locale' => $locale,
+            ],
+            [
+                'limit' => \PDO::PARAM_INT,
+                'offset' => \PDO::PARAM_INT,
+                'search' => \PDO::PARAM_STR,
+                'locale' => \PDO::PARAM_STR,
+            ],
```
Comment (mmetayer) [inline src/Akeneo/Pim/Structure/Bundle/Query/PublicApi/AttributeOption/Sql/SqlGetOptionsCountAndTranslationByAttribute.php:64]:
```suggestion
            [
                'select_attribute_types' => Connection::PARAM_STR_ARRAY,
            ],
```

File: src/Akeneo/Pim/Structure/Bundle/Query/PublicApi/AttributeOption/Sql/SqlGetOptionsCountAndTranslationByAttribute.php [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+        SQL;
+
+        $offset = \abs($page - 1) * $limit;
+
+        $rawResults = $this->connection->executeQuery(
+            $query,
+            [
+                'limit' => $limit,
+                'offset' => $offset,
+                'search' => strtolower($search ?? '') . '%',
```
Comment (mmetayer) [inline src/Akeneo/Pim/Structure/Bundle/Query/PublicApi/AttributeOption/Sql/SqlGetOptionsCountAndTranslationByAttribute.php:56]:
```suggestion
                'search' => \sprintf('%%%s%%', $search ?? ''),
                'select_attribute_types' => [AttributeTypes::OPTION_SIMPLE_SELECT, AttributeTypes::OPTION_MULTI_SELECT'],
```

File: src/Akeneo/Pim/Structure/Bundle/Query/PublicApi/AttributeOption/Sql/SqlGetOptionsCountAndTranslationByAttribute.php [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        SELECT a.code, COALESCE(t.label, a.code) AS label, count(o.id) as options_count
+        FROM pim_catalog_attribute a
+            LEFT JOIN pim_catalog_attribute_option o ON o.attribute_id = a.id
+            LEFT JOIN pim_catalog_attribute_translation t ON t.foreign_key = a.id
+                AND t.locale = :locale
+        WHERE attribute_type IN ('pim_catalog_simpleselect', 'pim_catalog_multiselect')
+        $querySearch
+        GROUP BY t.label, a.code
+        ORDER BY t.label, a.code
+        LIMIT :limit OFFSET :offset
```
Comment (mmetayer) [inline src/Akeneo/Pim/Structure/Bundle/Query/PublicApi/AttributeOption/Sql/SqlGetOptionsCountAndTranslationByAttribute.php:46]:
```suggestion
        SELECT a.code, t.label, COUNT(o.id) AS options_count
        FROM pim_catalog_attribute a
            LEFT JOIN pim_catalog_attribute_option o ON o.attribute_id = a.id
            LEFT JOIN pim_catalog_attribute_translation t ON t.foreign_key = a.id AND t.locale = :locale
        WHERE a.attribute_type IN (:select_attribute_types)
        AND COALESCE(t.label, a.code) LIKE :search
        GROUP BY a.code, t.label
        ORDER BY COALESCE(t.label, a.code)
        LIMIT :limit OFFSET :offset
```

Comment (mmetayer) [reply src/Akeneo/Pim/Structure/Bundle/Query/PublicApi/AttributeOption/Sql/SqlGetOptionsCountAndTranslationByAttribute.php:46]: - you don't want to return the code if the label is null, the client (i.e the UI most of the time) should handle this case itself
- no need to lower, the DB is case insensitive
- use constants for attribute types
- usually in the PIM, searches are performed with `LIKE '%search%'` (= contains), and not `LIKE 'search%'` (= starts with)

File: src/Akeneo/Pim/Structure/Bundle/Query/PublicApi/AttributeOption/Sql/SqlGetOptionsCountAndTranslationByAttribute.php [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+        SQL;
+
+        $offset = \abs($page - 1) * $limit;
```
Comment (mmetayer) [inline src/Akeneo/Pim/Structure/Bundle/Query/PublicApi/AttributeOption/Sql/SqlGetOptionsCountAndTranslationByAttribute.php:49]:
```suggestion
```

File: src/Akeneo/Pim/Structure/Bundle/Query/PublicApi/AttributeOption/Sql/SqlGetOptionsCountAndTranslationByAttribute.php (hunk 2 above)
Comment (amigaire) [reply src/Akeneo/Pim/Structure/Bundle/Query/PublicApi/AttributeOption/Sql/SqlGetOptionsCountAndTranslationByAttribute.php]: I kept all type declaration and added the 'select_attribute_types'. Without the type declarations, I had this error :

> AkeneoTest\Pim\Structure\Integration\AttributeOption\SqlGetOptionsCountAndTranslationByAttributeIntegration::test_it_returns_options_count_and_translation_by_attribute_code_with_limit_and_offset
> Doctrine\DBAL\Exception\SyntaxErrorException: An exception occurred while executing 'SELECT a.code, t.label, COUNT(o.id) AS options_count
> FROM pim_catalog_attribute a
>     LEFT JOIN pim_catalog_attribute_option o ON o.attribute_id = a.id
>     LEFT JOIN pim_catalog_attribute_translation t ON t.foreign_key = a.id AND t.locale = ?
> WHERE a.attribute_type IN (?, ?)
> AND COALESCE(t.label, a.code) LIKE ?
> GROUP BY a.code, t.label
> ORDER BY COALESCE(t.label, a.code)
> LIMIT ? OFFSET ?' with params ["en_US", "pim_catalog_simpleselect", "pim_catalog_multiselect", "%%", 1, 4]:
> 
> SQLSTATE[42000]: Syntax error or access violation: 1064 You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near ''1' OFFSET '4'' at line 9

File: tests/back/Pim/Structure/Integration/AttributeOption/SqlGetOptionsCountAndTranslationByAttributeIntegration.php [hunk 6]
```diff
@@ -0,0 +1,210 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of the Akeneo PIM Enterprise Edition.
```
Comment (pierallard) [inline tests/back/Pim/Structure/Integration/AttributeOption/SqlGetOptionsCountAndTranslationByAttributeIntegration.php:6]: This is not EE :)

**Review by pierallard — APPROVED** [review]

**Pull request merged**